### PR TITLE
Implement InitControl object for setting static controls before a run

### LINKED ISSUE #2913: Some MPI launchers call setsid(2) which disconnects the application session from the bash shell
**Describe the bug**
I tried to set MSRs inside of a PALS runscript and then call the integration experiment infrastructure. I expected my MSRs to be set and then the experiment to be launched. Instead, I got a geopm service write lock and the experiment failed.

**GEOPM version**
Output from `geopmread --version`
as root: 2.0.0+dev411g47f3f6e3
as user: 2.0.0+dev104g0351e9fe

**Error Message**
```

geopmd[31818]: The call io.github.geopm.PlatformWriteControl has failed with an exception:
geopmd[31818]: Traceback (most recent call last):
geopmd[31818]:   File "/usr/lib/python3.6/site-packages/dasbus/server/handler.py", line 459, in _method_callback
geopmd[31818]:     **additional_args
geopmd[31818]:   File "/usr/lib/python3.6/site-packages/dasbus/server/handler.py", line 265, in _handle_call
geopmd[31818]:     return handler(*parameters, **additional_args)
geopmd[31818]:   File "/usr/lib/python3.6/site-packages/geopmdpy/service.py", line 861, in PlatformWriteControl
geopmd[31818]:     self._platform.write_control(self._get_pid(**call_info), control_name, domain, domain_idx, setting)
geopmd[31818]:   File "/usr/lib/python3.6/site-packages/geopmdpy/service.py", line 675, in write_control
geopmd[31818]:     self._write_mode(client_pid)
geopmd[31818]:   File "/usr/lib/python3.6/site-packages/geopmdpy/service.py", line 711, in _write_mode
geopmd[31818]:     raise RuntimeError(f'The PID {client_pid} requested write access, but the geopm service already has write mode client with PID or SID of {lock_pid}')
geopmd[31818]: RuntimeError: The PID 34005 requested write access, but the geopm service already has write mode client with PID or SID of 31667

```

### LINKED ISSUE #2918: As user of geopmlaunch I would like to statically write control values prior to the job being started
**Is your feature request related to a problem? Please describe.**
Issue: #2913 
I'm always frustrated when I try to write a control prior to `geopmlaunch` and run into the issue of write-lock contention.

**Describe the solution you'd like**
Input file encompassing the desired control settings for the run that can be consumed by the controller at startup.

OPEN: file format. JSON? Space-delimited text file?
Decision:
- Flat file
- One line per control/setting
- Consumed by Controller
- Need to be careful about parsing it, write comprehensive unit tests.

**Describe alternatives you've considered**
Using command-line arguments (might be complex).

Tasks:
- [x] #2919
- [x] #2920
- [x] #2921
- [x] #2922
- [x] #2923

### LINKED ISSUE #2919: Class for managing `geopm-write` file: parse file, generate list of request objects, do writes when requested
> estimate 1

Done criteria:
- [ ] Environment class modified to comprehend GEOPM_WRITE file
- [ ] New class (e.g. InitControl or similar) written to query Environment class for the file and perform the writes
- [ ] Debug build feature: Print the writes

### LINKED ISSUE #2920: Launcher command-line option for `geopm-write` option
> estimate 1

Done criteria:
- [ ] New argparse option
- [ ] Constructor args modified to add file for integration testing
- [ ] Documentation updates

### LINKED ISSUE #2921: Unit tests for new static settings (init/write control) class
> estimate 1

Done criteria:
- [ ] New unit tests for new object
- [ ] New unit tests for Environment class

### LINKED ISSUE #2922: Integration test for `geopm-write` option/functionality (make sure command-line options get passed)
> estimate 1

Done criteria:
- [ ] 2 new tests: 1 that runs the monitor with some static controls, 1 that runs an agent that sets controls where the static controls are the same as what the agent tries to do.

### LINKED ISSUE #2923: `gepom-write` option documentation (update launcher man page, environment variables)
> estimate 1

### PR DESCRIPTION
**Pull request by bgeltz**
- Fixes #2918.
- Fixes #2919.
- Fixes #2920.
- Fixes #2921.
- Fixes #2922.
- Fixes #2923.
- Fixes #2913.

### CODE REVIEW

File: src/Controller.cpp [hunk 1]
```diff
@@ -176,7 +177,8 @@ namespace geopm
                      nullptr,
                      environment().endpoint(),
                      environment().do_endpoint(),
-                     ApplicationSampler::default_shmkey())
+                     ApplicationSampler::default_shmkey(),
+                     std::unique_ptr<InitControl>(new InitControlImp(environment().init_control())))
```
Comment (cmcantalupo) [inline src/Controller.cpp]:
```suggestion
                     InitControl::make_unique(environment().init_control())
```

File: src/Controller.cpp [hunk 2]
```diff
@@ -199,7 +201,8 @@ namespace geopm
                            std::unique_ptr<EndpointUser> endpoint,
                            const std::string &endpoint_path,
                            bool do_endpoint,
-                           const std::string &shm_key)
+                           const std::string &shm_key,
+                           std::unique_ptr<InitControl> init_control)
```
Comment (cmcantalupo) [inline src/Controller.cpp]:
```suggestion
                           std::shared_ptr<InitControl> init_control)
```

File: src/Controller.cpp [hunk 3]
```diff
@@ -226,6 +229,7 @@ namespace geopm
         , m_do_endpoint(do_endpoint)
         , m_do_policy(do_policy)
         , m_shm_key(shm_key)
+        , m_init_control(std::move(init_control))
```
Comment (cmcantalupo) [inline src/Controller.cpp]: if it is a shared pointer, you can avoid the move

File: src/Controller.hpp [hunk 4]
```diff
@@ -163,6 +165,8 @@ namespace geopm
             std::vector<std::string> m_agent_policy_names;
             std::vector<std::string> m_agent_sample_names;
             std::string m_shm_key;
+
+            std::unique_ptr<InitControl> m_init_control;
```
Comment (cmcantalupo) [inline src/Controller.hpp]:
```suggestion
            std::shared_ptr<InitControl> m_init_control;
```

File: src/InitControl.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            InitControlImp(const InitControlImp &other) = delete;
+            InitControlImp &operator=(const InitControlImp &other) = delete;
+
+            InitControlImp(const std::string &input_file);
+            InitControlImp(const std::string &input_file,
+                           PlatformIO &platform_io);
+            virtual ~InitControlImp() = default;
+
+        private:
+            std::vector<std::tuple<std::string, std::string, int, double> >
```
Comment (cmcantalupo) [inline src/InitControl.hpp]: I'd kind of prefer a class local structure

```
struct m_request {std::string name, std::string domain, int index, double setting}; 
```
to a tuple. I think, as the code is written, this would only require a change to the function signature and the type of the return value.

File: src/InitControl.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#include <vector>
+#include <tuple>
+
+namespace geopm
+{
+    class InitControl
+    {
+        public:
+            InitControl() = default;
+            virtual ~InitControl() = default;
```
Comment (cmcantalupo) [inline src/InitControl.hpp:22]: I recommend adding at least factory method and a method that executes the write

```
static std::unique_ptr<InitControl> make_unique(const std::string &input_file);
void write_controls(void);
```

I think better still would be:

```
static std::unique_ptr<InitControl> make_unique();
void parse_input(const std::string &input_file);
void write_controls(void);
```

File: test/InitControlTest.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+TEST_F(InitControlTest, parse_valid_file)
+{
+    std::string contents = "# This is a comment\n"
+                           "FAKE_CONTROL0 board 0 123     # Test comment 0\n"
+                           "FAKE_CONTROL1 package 1 -7.77 # Test comment 1\n"
+                           "\n"
+                           "# This is another comment\n"
+                           "FAKE_CONTROL2 gpu 3 0         # Test comment 2\n";
```
Comment (cmcantalupo) [inline test/InitControlTest.cpp]: What about the strings

```
CPU_POWER_LIMIT package 0 2 00
CPU_POWER_LIMIT package 0 2#00
    #CPU_POWER_LIMIT package 0 100
CPU_POWER_LIMIT package 0 200 CPU_POWER_LIMIT package 1 150
```

**Review by cmcantalupo — COMMENTED** [review]
Looks good.  Some suggestions inline

File: src/InitControl.hpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+{
+    class InitControl
+    {
+        public:
+            struct m_request_s {
+                std::string name;
+                int domain;
+                int domain_idx;
+                double setting;
+            };
```
Comment (avilcheslopez) [inline src/InitControl.hpp]: Does this struct really need to be public? It's not used to interact with the interface and it's only used internally by the implementation (`InitControlImp`).

File: src/InitControl.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                    continue;
+                }
+                else if (line[comment_pos] != '#') {
+                    throw Exception("Invalid line comment or missing fields while parsing: " + line,
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+            else { // Match; Continue to try parsing the setting
+                if (match_line.suffix().str().size() > 0 && !string_begins_with(match_line.suffix(), "#")) {
+                    throw Exception("Data found in comment field while parsing: " + line,
```
Comment (avilcheslopez) [inline src/InitControl.cpp]: Minor note:

This error message might be a bit confusing. If I'm following correctly, this would get triggered when there's additional text after the portion that was matched, for example:

`CPU_POWER_LIMIT_CONTROL board 0 200   this was supposed to be a comment, but I forgot the #`

The error message would make me think that I included part of the settings within a comment, or that maybe there's something wrong with the domain, domain index or value I provided.

File: src/InitControl.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                                    + ": parsing: " + line,
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+        }
+    }
+
+    void InitControlImp::write_controls(void) const
+    {
+        for (const auto &[ name, domain, index, setting ] : m_requests) {
```
Comment (avilcheslopez) [inline src/InitControl.cpp]: nit:
```suggestion
        for (const auto &[name, domain, index, setting] : m_requests) {
```

File: test/InitControlTest.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    EXPECT_CALL(m_platform_io,
+                write_control("FAKE_CONTROL0", PlatformTopo::domain_name_to_type("board"), 0, 123));
+    EXPECT_CALL(m_platform_io,
+                write_control("FAKE_CONTROL1", PlatformTopo::domain_name_to_type("package"), 1, -7.77))
+        .WillOnce(Throw(geopm::Exception("Test-injected failure", GEOPM_ERROR_INVALID, __FILE__, __LINE__)));
+
+    m_init_control = std::make_shared<InitControlImp>(m_file_name, m_platform_io);
+    GEOPM_EXPECT_THROW_MESSAGE(m_init_control->write_controls(),
+                               GEOPM_ERROR_INVALID, "Test-injected failure");
+}
```
Comment (avilcheslopez) [inline test/InitControlTest.cpp:188]: Add a test that specifies the expected behavior when an empty file name is provided to the constructor

File: src/InitControl.cpp (hunk 9 above)
Comment (bgeltz) [reply src/InitControl.cpp]: Syntax error

File: src/InitControl.cpp [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+                                   PlatformIO &platform_io)
+        : m_platform_io(platform_io)
+    {
+        if (input_file != "") {
+            parse_input(input_file);
+        }
+    }
+
+    void InitControlImp::parse_input(const std::string &input_file)
+    {
```
Comment (bgeltz) [inline src/InitControl.cpp:41]: Warn if the file had no contents.

File: src/Controller.cpp [hunk 13]
```diff
@@ -176,7 +177,8 @@ namespace geopm
                      nullptr,
                      environment().endpoint(),
                      environment().do_endpoint(),
-                     ApplicationSampler::default_shmkey())
+                     ApplicationSampler::default_shmkey(),
+                     InitControl::make_unique(environment().init_control()))
```
Comment (bgeltz) [inline src/Controller.cpp]: Switch this around so:
1. InitControl is always created.  Remove the input file string from the factory method.
2. Add a method to the Environment singleton for `do_init_control`.  If it is_set, then call parse_input with the input filename string.
3. In all cases, call write_controls().  It will be a no-op when do_init_control was false.

File: src/InitControl.cpp [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        // [eE]?  Optional e or E
+        // [+-]?  Optional + or -
+        // [0-9]+ One or more digits
+        // )?     End non-capture group, optional
+        // )      End capture group
+        std::regex sci_notation_regex("([+-]?[0-9]*[.]?[0-9]+(?:[eE]?[+-]?[0-9]+)?)");
+
+        // ^\\s*  String begins with zero or more whitespace characters
+        // (\\w+) Capture group #1 for one or more alphanumeric characters (CONTROL NAME)
+        // \\s+   Zero or more whitespace characters
```
Comment (bgeltz) [inline src/InitControl.cpp]:
```suggestion
        // \\s+   One or more whitespace characters
```
Same for the others below

**Review by avilcheslopez — APPROVED** [review]
